benchmark/tools/compare.py

```diff
@@ -0,0 +1,261 @@
+                key_to_str(dict_key), context
+            ),
+            file=sys.stderr,
+        )
+    for dict_key in comparison_only:
+        print(
+            "WARNING: Key {} found in comparison only in context {}".format(
+                key_to_str(dict_key), context
+            ),
+            file=sys.stderr,
+        )
+
+
+def ratio(baseline: int | float, comparison: int | float) -> float:
+    """Compares the ratio between baseline and comparison. For runtimes, this is the speedup."""
+    return baseline / comparison
+
+
+def compare_benchmark(baseline: dict, comparison: dict, context: str):
+    """Compares a handful of keys and component breakdowns recursively, writing them with a suffix to the output"""
+    warn_on_inconsistent_keys(baseline, comparison, context)
+    result = {}
+    for key in baseline.keys():
+        sub_context = "{}.{}".format(context, key)
+        if key == "components":
+            assert isinstance(baseline[key], dict)
+            assert isinstance(comparison[key], dict)
+            warn_on_inconsistent_keys(baseline[key], comparison[key], sub_context)
+            result[key + suffix] = {
+                sub_key: ratio(baseline[key][sub_key], comparison[key][sub_key])
+                for sub_key in baseline[key]
+            }
+        elif isinstance(baseline[key], dict):
+            result[key] = compare_benchmark(baseline[key], comparison[key], sub_context)
+        elif key in comparison_keys:
+            result[key + suffix] = ratio(baseline[key], comparison[key])
+    return result
+
+
+def compare(baseline: dict, comparison: dict, context: str) -> dict:
+    """Compares a test case, keeping root-level values and recursing into benchmarks"""
+    warn_on_inconsistent_keys(baseline, comparison, context)
+    result = {}
+    for key in baseline.keys():
+        # we don't have lists on the test case root level
+        assert not isinstance(baseline[key], list)
+        if isinstance(baseline[key], dict):
+            benchmark_result = {}
+            warn_on_inconsistent_keys(
+                baseline[key], comparison[key], "{}.{}".format(context, key)
+            )
+            for benchmark_name in baseline[key].keys():
+                if isinstance(baseline[key][benchmark_name], dict):
+                    comparison_result = compare_benchmark(
+                        baseline[key][benchmark_name],
+                        comparison[key][benchmark_name],
+                        "{}.{}.{}".format(context, key, benchmark_name),
+                    )
+                    if len(comparison_result) > 0:
+                        benchmark_result[benchmark_name] = comparison_result
+            if len(benchmark_result) > 0:
+                result[key] = benchmark_result
+        else:
+            # everything that's not a dict should only depend on the key in the root level
+            if baseline[key] != comparison[key]:
```

Comment:
Not necessarily, these can also be things like `rows`, `columns` and `nonzeros`, which should be the same but don't need to be part of the keys (and I would like to keep them out of there, because inconsistencies in these values point to mismatching inputs)